File: 16-3sum-closest/16-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    nums=nums.sort((a,b)=>a-b);
    let answer=Infinity;
    for(let i=0;i<nums.length-2;i++){
        let left=i+1;
        let right=nums.length-1;
        while(left<right){
            const sum = nums[i]+nums[left]+nums[right];
            if(Math.abs(sum-target)<Math.abs(answer-target)) answer=sum;
            if(sum>target){
                right--;
            }else{
                left++;
            }
        } 
    }
    return answer;
};

File: 16-3sum-closest/16-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        length = len(nums)
        # if length==3:
        #     return sum(nums)
        nums.sort()
        total = nums[0]+nums[1]+nums[-1]
        
        for i in range(length-2):
            left = i+1
            right = length-1
            while left<right:
                    total_pointer = nums[i]+nums[left]+nums[right]
                    
                    if total_pointer > target : 
                        right -=1
                    elif total_pointer<target:
                        left+=1
                    else:
                        return total_pointer
                    
                    if abs(target - total_pointer)<abs(target - total):
                        total = total_pointer
                        
        return total
                    